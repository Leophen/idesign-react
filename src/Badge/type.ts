export interface BadgeProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 徽标内容，可为任意内容
   * @default 0
   */
  count?: React.ReactNode;
  /**
   * 封顶的数字值
   * @default 99
   */
  maxCount?: number;
  /**
   * 徽标颜色
   */
  color?: string;
  /**
   * 是否为红点
   * @default false
   */
  dot?: boolean;
  /**
   * 尺寸
   * @default medium
   */
  size?: 'small' | 'medium';
  /**
   * 徽标形状
   * @default circle
   */
  shape?: 'circle' | 'round';
  /**
   * 当数值为 0 时，是否展示徽标
   * @default true
   */
  showZero?: boolean;
  /**
   * 徽标偏移量
   */
  offset?: Array<string | number>;
}
