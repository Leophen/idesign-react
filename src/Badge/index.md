---
nav:
  title: 组件
  path: /components
group:
  title: 数据展示组件
  order: 4
order: 5
---

# Badge 提示徽章

提示徽章一般出现在图标或文字的右上角，以提供及时、重要的信息提示。

## 基本用法

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge>
        <Button>按钮</Button>
      </Badge>
      <Badge count="new">
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

## 最大数字

可通过 `maxCount` 属性设置最大数字，默认为 `99`，超出则加上 `+`。

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge count={10} maxCount={9}>
        <Button>按钮</Button>
      </Badge>
      <Badge count={100}>
        <Button>按钮</Button>
      </Badge>
      <Badge count={1000} maxCount={999}>
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

## 自定义颜色

可通过 `color` 属性自定义徽标颜色。

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge color="#7EBF50">
        <Button>按钮</Button>
      </Badge>
      <Badge color="#F2AD4E">
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

## 小红点徽章

可通过 `dot` 属性控制为小红点样式。

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge dot>
        <Button>按钮</Button>
      </Badge>
      <Badge dot count="new">
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

## 自定义尺寸

可通过 `size` 属性自定义徽标尺寸。

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge size="medium">
        <Button>按钮</Button>
      </Badge>
      <Badge size="small">
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

## 自定义形状

可通过 `shape` 属性自定义徽标形状。

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge shape="circle">
        <Button>按钮</Button>
      </Badge>
      <Badge shape="round">
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

## 徽标位置偏移

可通过 `offset` 属性设置徽标位置偏移量。

```tsx
import React from 'react';
import { Button, Badge } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Badge count={6}>
        <Button>按钮</Button>
      </Badge>
      <Badge count={6} offset={[10, 10]}>
        <Button>按钮</Button>
      </Badge>
      <Badge count={6} offset={[-10, 10]}>
        <Button>按钮</Button>
      </Badge>
    </div>
  );
};

export default App;
```

<API />
