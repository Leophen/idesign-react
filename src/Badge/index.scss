.i-badge {
  position: relative;
  display: inline-block;
  vertical-align: middle;
}

.i-badge__content {
  position: absolute;
  top: 0;
  right: 0;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  color: #fff;
  font-size: 12px;
  white-space: nowrap;
  text-align: center;
  background-color: var(--i-error);
  transform: translate(50%, -50%);
  transform-origin: 100% 0;
  &.i-badge--circle,
  &.i-badge--round {
    min-width: 20px;
    height: 20px;
    padding: 0 6px;
    line-height: 20px;
  }
  &.i-badge--circle {
    border-radius: 10px;
  }
  &.i-badge--round {
    border-radius: 6px;
  }
  &.i-badge__size-small {
    &.i-badge--circle,
    &.i-badge--round {
      min-width: 16px;
      height: 16px;
      padding: 0 4px;
      line-height: 16px;
    }
  }
  &.i-badge--dot {
    right: 1px;
    width: 6px;
    height: 6px;
    margin-top: 1px;
    border-radius: 50%;
  }
}
