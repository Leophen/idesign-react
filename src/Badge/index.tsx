import React from 'react';
import classNames from 'classnames';
import './index.scss';
import { BadgeProps } from './type';

const Badge: React.FC<BadgeProps> = (props) => {
  const {
    children = '',
    className,
    style,
    count = 0,
    maxCount = 99,
    color,
    dot = false,
    size = "medium",
    shape = "circle",
    showZero = true,
    offset,
    ...restProps
  } = props;

  const getDisplayCount = () => {
    if (typeof count === 'number' && count > maxCount) {
      return `${maxCount}+`;
    }
    return count;
  };

  let isHidden = !count;
  if (typeof count === 'number') {
    isHidden = count < 1 && !showZero;
  }

  const getOffset = () => {
    const result: React.CSSProperties = {}
    if (offset) {
      offset[0] && (result.right = +offset[0])
      offset[1] && (result.marginTop = +offset[1])
    }
    return result
  }

  return (
    <div
      className={classNames(
        'i-badge',
        className
      )}
      {...restProps}
    >
      {children}
      {!isHidden && (
        <sup
          className={classNames(
            'i-badge__content',
            dot ? `i-badge--dot` : shape && `i-badge--${shape}`,
            size === 'small' && `i-badge__size-${size}`
          )}
          style={{ ...style, background: color, ...getOffset() }}
        >
          {!dot ? getDisplayCount() : null}
        </sup>
      )}
    </div>
  );
};

Badge.displayName = 'Badge';

export default Badge;
