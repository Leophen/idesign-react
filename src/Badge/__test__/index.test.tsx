import { render } from '@testing-library/react';
import Badge from '../index';

describe('Badge 组件测试', () => {
  function renderSup(badge) {
    const { container } = render(badge);
    return container.firstChild.lastChild;
  }

  it('count', () => {
    expect(renderSup(<Badge />)).toHaveTextContent('0');
    expect(renderSup(<Badge count='new' />)).toHaveTextContent('new');
  });

  it('max count', () => {
    expect(renderSup(<Badge count={100} />)).toHaveTextContent('99+');
    expect(renderSup(<Badge count={10} maxCount={9} />)).toHaveTextContent('9+');
  });

  it('color', () => {
    const { asFragment } = render(<Badge color="red" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('dot', () => {
    expect(renderSup(<Badge dot />)).toHaveClass('i-badge--dot');
  })

  it('size', () => {
    expect(renderSup(<Badge size="small" />)).toHaveClass('i-badge__size-small');
  })

  it('shape', () => {
    expect(renderSup(<Badge shape="round" />)).toHaveClass('i-badge--round');
  })

  it('offset', () => {
    expect(renderSup(<Badge offset={[10, 10]} />)).toHaveStyle({ right: '10px', 'margin-top': '10px' });
  })
});

