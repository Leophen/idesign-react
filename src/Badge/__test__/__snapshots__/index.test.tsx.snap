// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge 组件测试 color 1`] = `
<DocumentFragment>
  <div
    class="i-badge"
  >
    <sup
      class="i-badge__content i-badge--circle"
      style="background: red;"
    >
      0
    </sup>
  </div>
</DocumentFragment>
`;
